### PR TITLE
Improve metadata perf and refactor ilasm/ildasm metadata usage

### PR DESCRIPTION
**Pull request by davidwrighton**
This change currently conflates 2 change
1. Removing function pointer indirections around various enumerator calls into metadata
2. Stop depending on coreclr shared library from ildasm/ilasm. Instead those binaries now are statically linked against all of the metadata logic

### CODE REVIEW

File: src/ilasm/CMakeLists.txt [hunk 1]
```diff
@@ -8,7 +8,6 @@ add_definitions(-D__ILASM__)
 add_definitions(-DFEATURE_CORECLR)
 
 include_directories(.)
-include_directories(../ildasm/unixcoreclrloader)
```
Comment (jkotas) [inline src/ilasm/CMakeLists.txt]: Delete everything in unixcoreclrloader directory?

File: src/ildasm/dasm.cpp [hunk 2]
```diff
@@ -308,32 +307,17 @@ extern CQuickBytes *        g_szBuf_JUMPPT;
 extern CQuickBytes *        g_szBuf_UnquotedProperName;
 extern CQuickBytes *        g_szBuf_ProperName;
 
-#ifdef FEATURE_PAL
-CoreCLRLoader *g_loader;
-#endif
 MetaDataGetDispenserFunc metaDataGetDispenser;
```
Comment (jkotas) [inline src/ildasm/dasm.cpp]: Nit: These vars are not needed anymore. Can just call directly.

File: src/ilasm/CMakeLists.txt (hunk 1 above)
Comment (davidwrighton) [reply src/ilasm/CMakeLists.txt]: Yeah, I need to do that before checkin, as well as change the IMDInternalImport guid, so nobody gets burned with invalid vtable dispatches on this change if they do something like use a dbi with a non-compatible dac. This is still WIP, as while I managed to get the Windows build to be workable, the Linux/Mac build is broken.

File: src/ilasm/CMakeLists.txt [hunk 3]
```diff
@@ -66,14 +77,47 @@ set(ILASM_LINK_LIBRARIES
...
+        uuid.lib
+        user32.lib
+        version.lib
+        shlwapi.lib
+        bcrypt.lib
+        RuntimeObject.lib
+    )
+else()
+    list(APPEND ILASM_LINK_LIBRARIES
+        ${START_WHOLE_ARCHIVE} # force all PAL objects to be included so all exports are available
```
Comment (janvorli) [inline src/ilasm/CMakeLists.txt]: This ${START_WHOLE_ARCHIVE} should not be needed. Only the coreclr needs this as it exports symbols, some of which are from PAL.

File: src/ildasm/exe/CMakeLists.txt [hunk 4]
```diff
@@ -49,21 +48,64 @@ if (WIN32)
...
+        uuid.lib
+        user32.lib
+        version.lib
+        shlwapi.lib
+        bcrypt.lib
+        RuntimeObject.lib
+    )
+else()
+    list(APPEND ILDASM_LINK_LIBRARIES
+        ${START_WHOLE_ARCHIVE} # force all PAL objects to be included so all exports are available
```
Comment (janvorli) [inline src/ildasm/exe/CMakeLists.txt]: Dtto here

File: src/md/staticmd/apis.cpp [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+#include "shimload.h"
+#include "metadataexports.h"
+#include "ex.h"
+#include "product_version.h"
+
+// ---------------------------------------------------------------------------
+// %%Function: MetaDataGetDispenser
+// This function gets the Dispenser interface given the CLSID and REFIID.
+// ---------------------------------------------------------------------------
+STDAPI DLLEXPORT MetaDataGetDispenser(            // Return HRESULT
```
Comment (davidwrighton) [inline src/md/staticmd/apis.cpp]: comment alignment

**Pull request merged**